src/ray/gcs/gcs_server/gcs_job_manager.cc

```diff
@@ -19,56 +19,101 @@
 namespace ray {
 namespace gcs {
 
+void GcsJobManager::Initialize(const GcsInitData &gcs_init_data) {
+  for (auto &item : gcs_init_data.Jobs()) {
+    auto job_data = std::make_shared<JobTableData>(item.second);
+    // Add job data to local cache.
+    jobs_.emplace(item.first, job_data);
+  }
+}
+
 void GcsJobManager::HandleAddJob(const rpc::AddJobRequest &request,
                                  rpc::AddJobReply *reply,
                                  rpc::SendReplyCallback send_reply_callback) {
   JobID job_id = JobID::FromBinary(request.data().job_id());
-  RAY_LOG(INFO) << "Adding job, job id = " << job_id
-                << ", driver pid = " << request.data().driver_pid();
-  auto on_done = [this, job_id, request, reply,
-                  send_reply_callback](const Status &status) {
-    if (!status.ok()) {
-      RAY_LOG(ERROR) << "Failed to add job, job id = " << job_id
-                     << ", driver pid = " << request.data().driver_pid();
-    } else {
-      RAY_CHECK_OK(gcs_pub_sub_->Publish(JOB_CHANNEL, job_id.Hex(),
-                                         request.data().SerializeAsString(), nullptr));
-      RAY_LOG(INFO) << "Finished adding job, job id = " << job_id
-                    << ", driver pid = " << request.data().driver_pid();
+  auto driver_pid = request.data().driver_pid();
+  RAY_LOG(INFO) << "Adding job, job id = " << job_id << ", driver pid = " << driver_pid
+                << ", config is:\n"
+                << request.data().config().DebugString();
+
+  std::shared_ptr<JobTableData> job_table_data;
+
+  auto iter = jobs_.find(job_id);
+  if (iter != jobs_.end()) {
+    // The job already exists, it must be submitted from the dashboard.
+    // The job state should be RUNNING or SUBMITTED.
+    const auto state = iter->second->state();
+    if (iter->second->timestamp() == request.data().timestamp() &&
+        state == rpc::JobTableData::RUNNING) {
+      // It is a duplicated message, just reply ok.
+      GCS_RPC_SEND_REPLY(send_reply_callback, reply, Status::OK());
+      return;
+    } else if (state != rpc::JobTableData::SUBMITTED) {
+      std::ostringstream ostr;
+      ostr << "Failed to add job " << job_id
+           << " as job id conflicts or state is unexpected.";
+      RAY_LOG(WARNING) << ostr.str();
+      GCS_RPC_SEND_REPLY(send_reply_callback, reply, Status::Invalid(ostr.str()));
+      return;
     }
+    RAY_CHECK(state == rpc::JobTableData::RUNNING ||
+              state == rpc::JobTableData::SUBMITTED);
+    job_table_data = iter->second;
+  } else {
+    job_table_data = std::make_shared<JobTableData>();
+  }
+
+  // Reserve these fields that may be filled when submitting.
+  auto is_submitted_from_dashboard = job_table_data->is_submitted_from_dashboard();
+  auto job_payload = job_table_data->job_payload();
+
+  // Just use the job_table_data come from raylet.
+  job_table_data->CopyFrom(request.data());
```

Comment:
I think this will be cleaner:
```
if (iter != jobs_.end()) {
   // update some fields if needed.
} else {
   job_table_data->CopyFrom(request.data());
}
```

Reply:
Too many fields need to be updated. And It is easy to miss some fields.